File: packages/xyd-opencli2rust/src/handler.ts
// Handler emission (the handler.ts mirror from opencli2go): positional path args,
// IsSet-style guarded query/header population, flatten-vs-json body assembly —
// ending in runtime::run_request so the CliOverrides hooks see every request.

import type { Command } from '@xyd-js/opencli';

import { buildLeafModel, type FlagModel, type FlagType, type LeafModel } from './model';
import { snakeCase } from './naming';
import { indent } from './rslit';

const q = (s: string) => JSON.stringify(s);

function getOne(id: string, t: FlagType): string {
  switch (t) {
    case 'bool':
      return `m.get_one::<bool>(${q(id)})`;
    case 'int':
      return `m.get_one::<i64>(${q(id)})`;
    case 'float':
      return `m.get_one::<f64>(${q(id)})`;
    default:
      return `m.get_one::<String>(${q(id)})`; // string | json | file
  }
}

/** The `String`-typed rendering of a read flag value, for query/header params. */
function stringValue(t: FlagType): string {
  return t === 'string' || t === 'json' || t === 'file' ? 'v.clone()' : 'v.to_string()';
}

function pathParts(model: LeafModel): { fmt: string; args: string[]; usedVars: Set<string> } {
  const segs = model.path.split(/(\{[^}]+\})/).filter((s) => s !== '');
  let fmt = '';
  const args: string[] = [];
  const usedVars = new Set<string>();
  for (const seg of segs) {
    const m = seg.match(/^\{(.+)\}$/);
    if (m) {
      const a = model.pathArgs.find((p) => p.wireName === m[1]);
      fmt += '{}';
      if (a) {
        args.push(`runtime::path_escape(${a.varName})`);
        usedVars.add(a.varName);
      } else {
        args.push('""');
      }
    } else {
      fmt += seg.replace(/\{/g, '{{').replace(/\}/g, '}}');
    }
  }
  return { fmt, args, usedVars };
}

/** Guarded param population for query/header/cookie flags. */
function paramLines(flags: FlagModel[], target: string): string[] {
  const lines: string[] = [];
  for (const f of flags) {
    if (f.flagType === 'slice') {
      lines.push(`if let Some(vs) = m.get_many::<String>(${q(f.flagName)}) {`);
      lines.push(`    ${target}.push((${q(f.wireName)}, vs.cloned().collect::<Vec<_>>().join(",")));`);
      lines.push('}');
    } else {
      lines.push(`if let Some(v) = ${getOne(f.flagName, f.flagType)} {`);
      lines.push(`    ${target}.push((${q(f.wireName)}, ${stringValue(f.flagType)}));`);
      lines.push('}');
    }
  }
  return lines;
}

export interface RenderedHandler {
  name: string;
  code: string;
}

/** Render the async handler fn for a leaf command from its x-openapi binding. */
export function renderHandler(pathNames: string[], command: Command): RenderedHandler {
  const model = buildLeafModel(command);
  // Snake the WHOLE joined name (not per segment): a keyword segment like
  // "match" needs no trailing-underscore guard inside a longer identifier,
  // and per-segment guards would produce non-snake-case `__` doubles.
  const name = snakeCase(['handle', ...pathNames].join('-'));

  const lines: string[] = [];
  const push = (s = '') => lines.push(s);

  // Path params (positional args) — only read the ones the path actually uses.
  const { fmt, args, usedVars } = pathParts(model);
  for (const a of model.pathArgs) {
    if (usedVars.has(a.varName)) {
      push(`let ${a.varName} = m.get_one::<String>(${q(a.argName)}).map(String::as_str).unwrap_or("");`);
    }
  }
  push(args.length ? `let path = format!(${q(fmt)}, ${args.join(', ')});` : `let path = ${q(model.path)}.to_string();`);

  // Query params.
  const queryFlags = model.flags.filter((f) => f.location === 'query');
  if (queryFlags.length) {
    push(`let mut query: Vec<(&'static str, String)> = Vec::new();`);
    for (const line of paramLines(queryFlags, 'query')) push(line);
  }

  // Header / cookie params.
  const headerFlags = model.flags.filter((f) => f.location === 'header' || f.location === 'cookie');
  if (headerFlags.length) {
    push(`let mut headers: Vec<(&'static str, String)> = Vec::new();`);
    for (const line of paramLines(headerFlags, 'headers')) push(line);
  }

  // Body.
  let bodyExpr: string | undefined;
  if (model.hasBody) {
    if (model.bodyStyle === 'json' && model.bodyJsonOption) {
      push(`let body_bytes = m.get_one::<String>(${q(model.bodyJsonOption)}).map(|v| v.clone().into_bytes());`);
      bodyExpr = 'body_bytes';
    } else {
      push('let mut body = serde_json::Map::new();');
      for (const f of model.flags.filter((x) => x.location === 'body')) {
        if (f.flagType === 'slice') {
          push(`if let Some(vs) = m.get_many::<String>(${q(f.flagName)}) {`);
          push(
            `    body.insert(${q(f.wireName)}.to_string(), serde_json::Value::Array(vs.cloned().map(serde_json::Value::String).collect()));`,
          );
          push('}');
        } else if (f.flagType === 'json') {
          // Accept either JSON (objects/arrays) or a bare scalar: a string|object
          // union field is commonly given a plain value, and an unparseable value
          // must not abort the whole request.
          push(`if let Some(v) = m.get_one::<String>(${q(f.flagName)}) {`);
          push('    let value = serde_json::from_str::<serde_json::Value>(v)');
          push('        .unwrap_or_else(|_| serde_json::Value::String(v.clone()));');
          push(`    body.insert(${q(f.wireName)}.to_string(), value);`);
          push('}');
        } else {
          const value =
            f.flagType === 'int' || f.flagType === 'float'
              ? 'serde_json::Value::from(*v)'
              : f.flagType === 'bool'
                ? 'serde_json::Value::Bool(*v)'
                : 'serde_json::Value::String(v.clone())';
          push(`if let Some(v) = ${getOne(f.flagName, f.flagType)} {`);
          push(`    body.insert(${q(f.wireName)}.to_string(), ${value});`);
          push('}');
        }
      }
      push('let body_bytes = match serde_json::to_vec(&serde_json::Value::Object(body)) {');
      push('    Ok(bytes) => bytes,');
      push('    Err(err) => return o.print_error(cmd_path, &runtime::Error::Invalid(err.to_string())),');
      push('};');
      bodyExpr = 'Some(body_bytes)';
    }
  }

  // Assemble the request.
  push('let req = runtime::Request {');
  push(`    method: ${q(model.method)},`);
  push('    path,');
  push(queryFlags.length ? '    query,' : '    query: Vec::new(),');
  push(headerFlags.length ? '    headers,' : '    headers: Vec::new(),');
  push(bodyExpr ? `    body: ${bodyExpr},` : '    body: None,');
  push('};');
  push('runtime::run_request(ctx, o, cmd_path, req).await');

  const readsMatches = usedVars.size > 0 || model.flags.length > 0;
  const mParam = readsMatches ? 'm' : '_m';
  const code = `async fn ${name}<O: CliOverrides>(
    ctx: &Context,
    o: &O,
    cmd_path: &[String],
    ${mParam}: &ArgMatches,
) -> ExitCode {
${indent(lines.join('\n'))}
}`;
  return { name, code };
}

File: crates/xyd_opencli2rust/src/handler.rs
//! Handler emission — port of handler.ts. Path args, guarded query/header
//! population, flatten-vs-json body assembly, ending in runtime::run_request.

use std::collections::HashSet;

use serde_json::Value;

use crate::model::{build_leaf_model, FlagModel, FlagType, LeafModel};
use crate::naming::snake_case;
use crate::rslit::{indent, json_str};

fn q(s: &str) -> String {
    json_str(s)
}

fn get_one(id: &str, t: FlagType) -> String {
    match t {
        FlagType::Bool => format!("m.get_one::<bool>({})", q(id)),
        FlagType::Int => format!("m.get_one::<i64>({})", q(id)),
        FlagType::Float => format!("m.get_one::<f64>({})", q(id)),
        _ => format!("m.get_one::<String>({})", q(id)), // string | json | file
    }
}

fn string_value(t: FlagType) -> &'static str {
    match t {
        FlagType::String | FlagType::Json | FlagType::File => "v.clone()",
        _ => "v.to_string()",
    }
}

/// Split a path keeping `{param}` tokens (mirrors JS `split(/(\{[^}]+\})/)`).
fn split_path_tokens(path: &str) -> Vec<String> {
    let chars: Vec<char> = path.chars().collect();
    let mut out: Vec<String> = Vec::new();
    let mut lit = String::new();
    let mut i = 0;
    while i < chars.len() {
        if chars[i] == '{' {
            if let Some(close) = chars[i + 1..].iter().position(|&c| c == '}') {
                if close >= 1 {
                    if !lit.is_empty() {
                        out.push(std::mem::take(&mut lit));
                    }
                    let token: String = chars[i..=i + 1 + close].iter().collect();
                    out.push(token);
                    i = i + 1 + close + 1;
                    continue;
                }
            }
        }
        lit.push(chars[i]);
        i += 1;
    }
    if !lit.is_empty() {
        out.push(lit);
    }
    out
}

struct PathParts {
    fmt: String,
    args: Vec<String>,
    used_vars: HashSet<String>,
}

fn path_parts(model: &LeafModel) -> PathParts {
    let segs = split_path_tokens(&model.path);
    let mut fmt = String::new();
    let mut args: Vec<String> = Vec::new();
    let mut used_vars: HashSet<String> = HashSet::new();
    for seg in segs {
        let is_token = seg.starts_with('{') && seg.ends_with('}') && seg.len() >= 3;
        if is_token {
            let inner = &seg[1..seg.len() - 1];
            fmt.push_str("{}");
            if let Some(a) = model.path_args.iter().find(|p| p.wire_name == inner) {
                args.push(format!("runtime::path_escape({})", a.var_name));
                used_vars.insert(a.var_name.clone());
            } else {
                args.push("\"\"".to_string());
            }
        } else {
            fmt.push_str(&seg.replace('{', "{{").replace('}', "}}"));
        }
    }
    PathParts {
        fmt,
        args,
        used_vars,
    }
}

/// Guarded param population for query/header/cookie flags.
fn param_lines(flags: &[&FlagModel], target: &str) -> Vec<String> {
    let mut lines: Vec<String> = Vec::new();
    for f in flags {
        if f.flag_type == FlagType::Slice {
            lines.push(format!(
                "if let Some(vs) = m.get_many::<String>({}) {{",
                q(&f.flag_name)
            ));
            lines.push(format!(
                "    {target}.push(({}, vs.cloned().collect::<Vec<_>>().join(\",\")));",
                q(&f.wire_name)
            ));
            lines.push("}".to_string());
        } else {
            lines.push(format!(
                "if let Some(v) = {} {{",
                get_one(&f.flag_name, f.flag_type)
            ));
            lines.push(format!(
                "    {target}.push(({}, {}));",
                q(&f.wire_name),
                string_value(f.flag_type)
            ));
            lines.push("}".to_string());
        }
    }
    lines
}

pub struct RenderedHandler {
    pub name: String,
    pub code: String,
}

/// Render the async handler fn for a leaf command from its x-openapi binding.
pub fn render_handler(path_names: &[String], command: &Value) -> RenderedHandler {
    let model = build_leaf_model(command);
    let mut joined = vec!["handle".to_string()];
    joined.extend(path_names.iter().cloned());
    let name = snake_case(&joined.join("-"));

    let mut lines: Vec<String> = Vec::new();

    // Path params — only read the ones the path actually uses.
    let PathParts {
        fmt,
        args,
        used_vars,
    } = path_parts(&model);
    for a in &model.path_args {
        if used_vars.contains(&a.var_name) {
            lines.push(format!(
                "let {} = m.get_one::<String>({}).map(String::as_str).unwrap_or(\"\");",
                a.var_name,
                q(&a.arg_name)
            ));
        }
    }
    if !args.is_empty() {
        lines.push(format!(
            "let path = format!({}, {});",
            q(&fmt),
            args.join(", ")
        ));
    } else {
        lines.push(format!("let path = {}.to_string();", q(&model.path)));
    }

    // Query params.
    let query_flags: Vec<&FlagModel> = model
        .flags
        .iter()
        .filter(|f| f.location == "query")
        .collect();
    if !query_flags.is_empty() {
        lines.push("let mut query: Vec<(&'static str, String)> = Vec::new();".to_string());
        lines.extend(param_lines(&query_flags, "query"));
    }

    // Header / cookie params.
    let header_flags: Vec<&FlagModel> = model
        .flags
        .iter()
        .filter(|f| f.location == "header" || f.location == "cookie")
        .collect();
    if !header_flags.is_empty() {
        lines.push("let mut headers: Vec<(&'static str, String)> = Vec::new();".to_string());
        lines.extend(param_lines(&header_flags, "headers"));
    }

    // Body.
    let mut body_expr: Option<String> = None;
    if model.has_body {
        if model.body_style.as_deref() == Some("json") && model.body_json_option.is_some() {
            let opt = model.body_json_option.as_ref().unwrap();
            lines.push(format!(
                "let body_bytes = m.get_one::<String>({}).map(|v| v.clone().into_bytes());",
                q(opt)
            ));
            body_expr = Some("body_bytes".to_string());
        } else {
            lines.push("let mut body = serde_json::Map::new();".to_string());
            for f in model.flags.iter().filter(|x| x.location == "body") {
                if f.flag_type == FlagType::Slice {
                    lines.push(format!(
                        "if let Some(vs) = m.get_many::<String>({}) {{",
                        q(&f.flag_name)
                    ));
                    lines.push(format!(
                        "    body.insert({}.to_string(), serde_json::Value::Array(vs.cloned().map(serde_json::Value::String).collect()));",
                        q(&f.wire_name)
                    ));
                    lines.push("}".to_string());
                } else if f.flag_type == FlagType::Json {
                    lines.push(format!(
                        "if let Some(v) = m.get_one::<String>({}) {{",
                        q(&f.flag_name)
                    ));
                    lines.push(
                        "    let value = serde_json::from_str::<serde_json::Value>(v)".to_string(),
                    );
                    lines.push(
                        "        .unwrap_or_else(|_| serde_json::Value::String(v.clone()));"
                            .to_string(),
                    );
                    lines.push(format!(
                        "    body.insert({}.to_string(), value);",
                        q(&f.wire_name)
                    ));
                    lines.push("}".to_string());
                } else {
                    let value = match f.flag_type {
                        FlagType::Int | FlagType::Float => "serde_json::Value::from(*v)",
                        FlagType::Bool => "serde_json::Value::Bool(*v)",
                        _ => "serde_json::Value::String(v.clone())",
                    };
                    lines.push(format!(
                        "if let Some(v) = {} {{",
                        get_one(&f.flag_name, f.flag_type)
                    ));
                    lines.push(format!(
                        "    body.insert({}.to_string(), {});",
                        q(&f.wire_name),
                        value
                    ));
                    lines.push("}".to_string());
                }
            }
            lines.push(
                "let body_bytes = match serde_json::to_vec(&serde_json::Value::Object(body)) {"
                    .to_string(),
            );
            lines.push("    Ok(bytes) => bytes,".to_string());
            lines.push("    Err(err) => return o.print_error(cmd_path, &runtime::Error::Invalid(err.to_string())),".to_string());
            lines.push("};".to_string());
            body_expr = Some("Some(body_bytes)".to_string());
        }
    }

    // Assemble the request.
    lines.push("let req = runtime::Request {".to_string());
    lines.push(format!("    method: {},", q(&model.method)));
    lines.push("    path,".to_string());
    lines.push(if !query_flags.is_empty() {
        "    query,".to_string()
    } else {
        "    query: Vec::new(),".to_string()
    });
    lines.push(if !header_flags.is_empty() {
        "    headers,".to_string()
    } else {
        "    headers: Vec::new(),".to_string()
    });
    lines.push(match &body_expr {
        Some(e) => format!("    body: {e},"),
        None => "    body: None,".to_string(),
    });
    lines.push("};".to_string());
    lines.push("runtime::run_request(ctx, o, cmd_path, req).await".to_string());

    let reads_matches = !used_vars.is_empty() || !model.flags.is_empty();
    let m_param = if reads_matches { "m" } else { "_m" };
    let code = format!(
        "async fn {name}<O: CliOverrides>(\n    ctx: &Context,\n    o: &O,\n    cmd_path: &[String],\n    {m_param}: &ArgMatches,\n) -> ExitCode {{\n{}\n}}",
        indent(&lines.join("\n"), 1)
    );
    RenderedHandler { name, code }
}
